[PR-1320] Article about untyped lambda calculus

lierdakil opened the pull request:
This should (hopefully) be ready for review, but I might give it another read and pick some nits later this week.

lierdakil committed:
[PR-1320] Article about untyped lambda calculus

Heimdell commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+[Kleene–Rosser paradox]: https://en.wikipedia.org/wiki/Kleene%E2%80%93Rosser_paradox
+
+The lambda symbol used in the standard notation, from which lambda calculus derives its name, most likely was chosen arbitrarily.
+However, there's an apocryphal story the notation was derived from $\hat x$, morphing first into $\Lambda x$, and later into $\lambda x$.
```
Because publisher had no `^` litera, yes :D

lierdakil replied:
I found no hard evidence for this, though, so this might be folklore.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+
+There are two broad kinds of lambda calculus, _typed_, and _untyped_.
+The _untyped_ kind, discussed in this article, expresses an unrestricted computation.
+However, as a system of logic, it is self-contradictory.
```
Could you put a link that describes why? Perhaps here there's something: https://serokell.io/blog/formal-verification-history

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+
+## The pure untyped lambda calculus
+
+The untyped lambda calculus is, first and foremost, a consistent equational theory of functions.
```
Most of it is followable, but this sentence sticks out as some words that only those that know, know. Perhaps add a rephrasing or something?

DanielRrr replied:
I don't like the term 'equational theory' since it might be confusing with universal algebra. I would consider 'system of expression rewriting' or something like that.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+The list of languages based on lambda calculus includes:
+
+- Haskell
+- Most languages in the ML family, like OCaml and F#
```
Haskell is also in the ML family, no? Somewhat odd phrasing.

DanielRrr replied:
Haskell doesn't belong to the ML family

NaeosPsy replied:
Fair enough. 😅

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+Understanding lambda calculus can be very helpful when talking about Haskell programs.
+It is also used extensively in programming language theory research and functional language design.
+
+There are two broad kinds of lambda calculus, _typed_, and _untyped_.
```
```suggestion
There are two broad kinds of lambda calculus: _untyped_ and _typed_.
```

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+- Haskell
+- Most languages in the ML family, like OCaml and F#
+- Cardano blockchain's Plutus Core
+
```
I would add dependently typed languages as well

DanielRrr replied:
And the Lisp family.

Perhaps, it's worth saying that mainstream languages mostly have lambdas as anonymous functions.

lierdakil replied:
It's complicated with Lisp. The "wisdom of the crowds" insists McCarthy based Lisp on λ-calculus, but the man himself insists this was not the case. I'm inclined to believe the latter, because while it's obvious McCarthy was _aware_ of λ-calculus, AFAIK the first versions of Lisp didn't really fit into any λ-calculus evaluation strategy. Modern Lisp dialects are arguably similar enough to λ-calculus to say they're loosely based on it, but historically it seems to be more of a case of convergent evolution. It is pretty hard to untangle what inspired what, though.

lierdakil replied:
As for "lambdas as anonymous functions", it's also kind of complicated. While the _syntax_ for anonymous functions is often loosely inspired by λ-calculus (often very indirectly), the actual semantics have little to do with it, so I don't know how useful it is to mention that.

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+The initial formulation had a logical inconsistency known as [Kleene–Rosser paradox].
+To sidestep this issue, Church isolated the part of lambda calculus relevant only to computation in 1936.
+This isolate is now known as the untyped lambda calculus.
+Later, In 1940, a typed version of lambda calculus was introduced, which has weaker expressive power but is logically consistent.
```
One may also mention that the idea of types has been initially proposed by Bertrand Russell (1903, 1907, 1911). The initial version of a typed lambda was a combination of the lambda calculus and Russell's type theory.

lierdakil replied:
I've added a mention that typed λ-calculus is based on Russel's type theory. Getting into it beyond that seems slightly off-topic. Hopefully I'll remember to get into more detail in the planned simply typed λ-calculus article.

DanielRrr reviewed (commented):
I'm gonna leave more comments a bit later

lierdakil committed:
fixup! [PR-1320] Article about untyped lambda calculus

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+In eager evaluation, arguments are evaluated before the function call.
+Call-by-name, on the other hand, corresponds to lazy evaluation.
+In lazy evaluation, the arguments potentially are never evaluated at all.
+
```
I think this part might require some examples to compare strategies and their pros and cons in the concrete case.

lierdakil replied:
It's a little hard to construct examples without first introducing some meaningful terms, which is done below. I could of course use impure calculus (e.g. I could introduce naturals), but doing that for a one-off example seems kind of extraneous.

I'll see what I can do with simple terms, but I'm not very optimistic frankly.

lierdakil replied:
Well, I've added _an_ example. It illustrates some usual points, but I can't say I'm very happy with how it turned out. I expect most readers will skip right over.

NaeosPsy replied:
Yeah, I don't think it is really helpful. 

If you use one in the article, I would introduce how that works, tell that there are options, perhaps a fun fact or two about Haskell, but wouldn't explain this all, it doesn't really help the reader.

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+### Multivariate functions
+
+Any multivariate function is the same as multiple nested functions returning other functions.
+If you're familiar with functional programming, you might know such functions are called curried functions, named in honor of Haskell Curry.
```
s/named in honor/named after

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+The arrow $\to$ represents one evaluation step:
+
+$(\lambda x\;y. x\;y\;fls)\;(\lambda a\; b. a)\;(\lambda c\; d. c)$
+$\to (\lambda y. (\lambda a\; b. a)\;y\;fls)\;(\lambda c\; d. c)$
```
I would use the $\to_{\beta}$ arrow

lierdakil replied:
I don't think I use arrows in any other sense in this article, so I don't think noisier notation is justified here.

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+$$X = \lambda f. \Omega\; (\lambda x. f\;(x\;x)).$$
+
+We should note that it's not the only fixed-point combinator.
+For instance, another famous one is
```
I would say explicitly that it's named after Turing

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+
+> ... one of the myths concerning LISP that people think up or invent for themselves becomes apparent, and that is that LISP is somehow a realization of the lambda calculus, or that was the intention.
+> The truth is that I didn't understand the lambda calculus, really.
+>
```
Don't believe him, lol

lierdakil replied:
As I outlined above, there is some corroborating evidence to his claim.

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,461 @@
+[brief-incomplete-history]: http://james-iry.blogspot.com/2009/05/brief-incomplete-and-mostly-wrong.html
+
+Lambda calculus, initially envisioned as a formal logic system, was developed by Alonzo Church around the 1930s to explore the foundations of mathematics.
+The initial formulation had a logical inconsistency known as [Kleene–Rosser paradox].
```
It's worth referring to something like this https://plato.stanford.edu/entries/paradoxes-contemporary-logic/#IncoCertFormLogi

DanielRrr reviewed (commented):
Looks OK. I have several tiny remarks.

lierdakil committed:
fixup! [PR-1320] Article about untyped lambda calculus

Co-authored-by: NaeosPsy <35335196+NaeosPsy@users.noreply.github.com>

lierdakil committed:
fixup! [PR-1320] Article about untyped lambda calculus

lierdakil committed:
fixup! [PR-1320] Article about untyped lambda calculus

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+>
+> -- John McCarthy, Lisp session, History of Programming Languages
+
+Also, while not strictly speaking based on lambda calculus, C++ template metalanguage is a functional language in its own right, so much of what we discuss below happens to be applicable.
```
Is this necessary? I feel like we can cut this.

lierdakil replied:
While not strictly speaking _necessary_ (nothing in the intro is strictly speaking necessary besides "λ-calculus is every FP programming language invented, ever" :laughing: ), I, for one, find this curious, and it suggests that lambda-calculus is applicable outside the hard-core FP bubble. If the point of the intro is to provide the motivation for the reader, I think this paragraph is useful. I'd rather trim down the LISP discussion if you think the intro is a bit too long.

NaeosPsy replied:
I see. In that case, perhaps we can just follow this comment: https://github.com/serokell/blog-posts/pull/12#discussion_r857395866

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+
+## The pure untyped lambda calculus
+
+The untyped lambda calculus is, first and foremost, a consistent expression rewriting system.
```
Would "consistent system for rewriting expressions" also be correct? I feel it would sound better.

lierdakil replied:
Technically, it's the same words in slightly different order, so it's not wrong. Can't say I like it, but that's why we have an editor, I guess :laughing:

Fixed.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+[Kleene–Rosser paradox]: https://en.wikipedia.org/wiki/Kleene%E2%80%93Rosser_paradox
+[paradoxes-contemporary-logic]: https://plato.stanford.edu/entries/paradoxes-contemporary-logic/#IncoCertFormLogi
+
+The lambda symbol used in the standard notation, from which lambda calculus derives its name, most likely was chosen arbitrarily.
```
This is a cool fact. But it feels that right here, a person that has not seen the syntax at all (since you introduce it later) will not find it relatable. Perhaps cut it or move it below?

lierdakil replied:
Good point, moved.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+This insight arguably kickstarted the research on functional programming languages.
+
+## The pure untyped lambda calculus
+
```
It feels that it would be nice to start with some justification for "pure" just so it flows a little bit better

E.g. something like this: 

The best example for starters is pure untyped lambda calculus, "pure" meaning that it only contains functions and variables.

This calculus is, first and foremost, a consistent system for rewriting expressions.

lierdakil replied:
> start with some justification for "pure"

To me, this feels weird. Specifically, why are we explaining what is "pure" without explaining what λ-calculus is? "Pure" is a qualifier on λ-calculus. It's like explaining a "pink flamingo" starting from "pink". It's a flamingo that happens to be pink, why are we starting with the least important bit? This feels backwards.

NaeosPsy replied:
Yeah, ok, makes sense. 

The main problem here was that there is no justification for introducing exactly "pure untyped lambda calculus"; there's only definitions. 

So you can also have the same sentence order as before, just I'd suggest to add something like: "the most basic version of this calculus is pure untyped lambda calculus, which contains "

lierdakil replied:
Well, here's the slightly updated version:

> The simplest, smallest type of lambda calculus is the pure untyped lambda calculus, so we'll start with it.

> The untyped lambda calculus is, first and foremost, a consistent system for rewriting expressions.
> Lambda calculus that only contains functions and variables is called pure, meaning it doesn't have anything beyond what's strictly necessary.

NaeosPsy replied:
This doesn't really sound good. 😅

Would you be fine if we switched the last two sentences to: 

Like any untyped lambda calculus, it is first and foremost a consistent system for rewriting expressions.
And it is called pure because it doesn't have anything beyond what's strictly necessary – it only contains functions and variables.

lierdakil replied:
Fine by me. Done.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+We can formally define the pure untyped lambda calculus in terms of _abstractions_ and _applications_.
+
+_Abstraction_, or functional abstraction, is a definition of a parametric expression, that is to say, a function.
+Lambda calculus only defines univariate functions, but that's easily extended to multivariate, as we'll see later.
```
I find "that's easily extended to multivariate" a bit confusing even though I understand what you mean.

Perhaps "it can be easily extended" as in lambda calculus, or those as in functions? Or some other rewording.

lierdakil replied:
Reworded to

> Lambda calculus only defines univariate functions, but we can easily extend it to include multivariate functions too, as we'll see later.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+
+We can formally define the pure untyped lambda calculus in terms of _abstractions_ and _applications_.
+
+_Abstraction_, or functional abstraction, is a definition of a parametric expression, that is to say, a function.
```
Is abstraction an action or an item? 

If it is an item, then it should be _an abstraction_ I think. If it is the action like application, then it seems that it would be the _the act of defining a parametric expression_. I think that would be clearer. 

Above as well, perhaps it should be "in terms of two actions: abstraction and application".

lierdakil replied:
Uh. No, abstraction is a term, it doesn't act on terms, that's a confusing word choice on my part. Rephrased to

> An _abstraction_, or a functional abstraction, is a parametric expression, that is to say, a function.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+However, we might need parenthesis around the whole abstraction itself.
+We will also eschew parentheses for nested abstractions, e.g. $\lambda x. \lambda y. x\;y$ is $\lambda x. (\lambda y. (x\;y)).$
+
+Variables in the body of an abstraction that are parameters of said abstraction are called _bound variables_.
```
Feels like free & bound variables & alpha-equivalence go with "abstraction", so they should probably be moved up? 

I believe you could make level-3 subheaders for abstraction (where is everything regarding variables) and application, it would look clearer like that.

lierdakil replied:
Good point, moved.

Although I'm not sure about the sub-headers. FWIW I've added them, but it feels kind of meh.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+Expressions that differ only in the names of bound variables are called _alpha-equivalent_.
+For example, $\lambda x. x$ and $\lambda y. y$ are alpha-equivalent, but $\lambda x. y$ and $\lambda y. z$ are not.
+
+## Defining the computation
```
```suggestion
## Defining computation in lambda calculus
```

lierdakil replied:
Ehh... I'm not sure about this change. See, the header is not wrong, we're actually defining _the_ computation by the merit of the fact that every possible computation is representable in terms of λ-calculus :confused: We could define it any other way (Turing machines, Post machines, μ-recursive functions, Markov's normalized algorithms, etc), but it'll still be the same thing, all definitions are mathematically equivalent.

NaeosPsy replied:
I believe the second would still agree with what you say. 

For me, saying "defining the computation" sounds wrong on a feels level. Could be because computation is a process, and those don't take "the"?

lierdakil replied:
I've split the difference and removed "the". Feel free to argue if you feel that "in lambda calculus" is important (but IMO it's implied anyway, and shorter, catchier titles = more better in my book)

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+
+## Defining the computation
+
+The step of the computation is called beta-reduction.
```
This sentence I don't really like. 

Perhaps something like: A step of computation in lambda calculus is called beta reduction.

lierdakil replied:
Now this is fine. Changed.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+In the left term, it refers to the bound variable; in the right one, it refers to a free variable.
+To avoid ambiguity, we can rename $y$ in the first expression to, say, $z$: $(\lambda x. \lambda z. x\;z)\;(\lambda x. x\;y).$
+Then we can do the rest of the steps with no fear of confusion: $\lambda z. (\lambda x. x\;y)\;z.$
+
```
Feel like this should be a subsection called evaluation strategies.

Heimdell replied:
No, that's "we don't use de Bruijn indices" section.

lierdakil replied:
@NaeosPsy I've added a level 3 header where it seemed appropriate (i.e. after "It brings us to the discussion of evaluation strategies."). The highlighted part is not it, though.

@Heimdell :unamused: I actually introduce de Bruijn encoding in the exercises section. It's nice when implementing λ-calculus in software, but not so well suited for napkin math we're doing in this here article.

BTW, you might want to check my explanation of de Bruijn encoding, I'm not sure if it's quite on-point because it's been a while since I implemented my λ-calculus evaluation engine.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+
+So, let us declare two combinators,
+$tru = \lambda x\; y. x,$ and $fls = \lambda x\;y. y.$
+We're calling these $tru$ and $fls$ to disambiguate them from atomic boolean values.
```
What is an atomic boolean value?

Heimdell replied:
`true` and `false` in "normal" languages?

NaeosPsy replied:
```suggestion
We're calling these $tru$ and $fls$ to disambiguate them from "regular" boolean values, which are atomic.
```

Something like this, perhaps.  It is clear what is meant from the surrounding context, but it can raise an unnecessary "huh", I think.

lierdakil replied:
"Atomic" is non-decomposable, i.e. it doesn't have any internal structure (hence very different from Church booleans that are actually abstractions). It's a pretty common expression in math circles, but I see your point, a casual reader might not be familiar. Tried to explain it better:

> We're calling these $tru$ and $fls$ to disambiguate them from actual boolean values $true$ and $false$ that aren't actually abstractions.
> $true$ and $false$ are, hopefully obviously, not a part of the pure lambda calculus, but we can introduce them as terms in an impure lambda calculus.

NaeosPsy replied:
Yeah, I understood it once I thought about it. Sounds good, thank you. 👍

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+### Church booleans
+
+Let us now introduce booleans.
+We will follow Church here.
```
Do you think it is clear what this sentence means? Like, I understood, but some might be confused that you mean that you will cover the Church encoding of booleans

Heimdell replied:
Or https://wiki.c2.com/?ChurchOfTheLeastFixedPoint.

lierdakil replied:
Changed to

> We will do it in the same way Church did in his original paper.

Although I was briefly tempted to add "the" before "Church" and call it a day :joy:

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+
+### Church pairs
+
+Since we have booleans now, we can also encode pairs (and by extension, tuples as trees or right combs) as functions, returning one value when given $tru$ as an argument, and another when given $fls$.
```
Not sure it is worth it to mention about tuples as trees and right combs.

lierdakil replied:
I can drop "as trees or right combs", that's a reasonably extraneous remark.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+$= (\lambda a\; b. a)\;x\;y$
+$\Rightarrow x.$
+
+Here, double arrow $\Rightarrow$ signifies more than one step of beta-reduction.
```
more than one step -> multiple steps of beta-reduction?

lierdakil replied:
Technically, "multiple" is ill-defined, as it can depending on the context mean "one or more" or "more than one", but in this case it doesn't really matter. Changed.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+### Church naturals
+
+Since we encoded both pairs and booleans, we already can encode naturals simply as sequences of booleans, that is, as binary numbers.
+In practice, this turns out to be cumbersome, however, so instead, we'll define naturals similar to the Peano construction.
```
Can you add some link to Peano numbers for the teapots in the audience, perhaps?

lierdakil replied:
Yep. Can't decide between [Wikipedia](https://en.wikipedia.org/wiki/Peano_axioms#Historical_second-order_formulation) and [Haskell Wiki](https://wiki.haskell.org/Peano_numbers). Haskell wiki entry is short and simple, but relies on the knowledge of Haskell. The Wikipedia article of course goes into much more (arguably, here, unnecessary) detail.

NaeosPsy replied:
I think Haskell Wiki is fine, Wikipedia one looks too scary for anybody who would need that link.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+
+It might be easier to see what's going on if we explicitly note the outer abstraction:
+
+$succ = \lambda n. \lambda s\; z. s\;(n\;s\;z).$
```
You use both $..$ and $$...$$, is that on purpose? 

It seems that $..$ renders much better for a text that uses a lot of math since $$...$$ gets centered which breaks the flow of the text too much. Perhaps we can just write everything with $..$?

lierdakil replied:
Here, we can, yes.

NaeosPsy commented on untyped-lambda-calculus/article.md:
Point at the end is the ending of the sentence? 

Since it renders in a new line, I don't feel it is necessary here and elsewhere.

lierdakil replied:
Eh, that's actually standard practice for formulae-heavy articles. It's harder to parse without puncutation.

NaeosPsy replied:
Ok, if that's the case, then I have no objections.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+We run into a bit of an issue when trying to define subtraction or a predecessor function.
+Turns out it's doable but neither pretty nor efficient.
+We'll use pairs of values containing two subsequent naturals.
+That way, we can stagger the counter by one.
```
Is  "stagger" specific terminology? I don't really understand the use of this word here.

lierdakil replied:
Okay, I may be misusing the word here, but I definitely encountered "staggered counter" in the literature (meaning a counter lagging by some specific offset). Rewrote to

> That way, we can have a counter that's lagging by one.

NaeosPsy replied:
Oh, ok. Yes, that's definitely clearer.

NaeosPsy replied:
You used the word later as well, btw, that should also probably be changed. (On line 361 for me)

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+
+$$pred = \lambda n. fst\; (n\; ss\; zz).$$
+
+With this definition, $pred\;c_0 = c_0,$ which isn't ideal.
```
I would put the comma outside the latex, It renders two different commas, which isn't very aesthetic, I think.

lierdakil replied:
The reason I put commas and other punctuation marks inside LaTeX is that otherwise those punctuation marks will eventually run away to the next line. Arguably, this is worse than a slight font mismatch:

![image](https://user-images.githubusercontent.com/7275622/165609458-40690c60-dc4d-45db-a9b1-73ea531b2a3a.png)

NaeosPsy replied:
It is indeed worse. Okay. 😁

NaeosPsy committed:
First editing read

NaeosPsy commented on untyped-lambda-calculus/article.md:
I would hide this under some Tips for handling name collisions spoiler or something, it puts too much text between the end of "real text" and more articles + subscriber box, which isn't aesthetic.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,493 @@
+Understanding lambda calculus can be very helpful when talking about Haskell programs.
+It is also used extensively in programming language theory research and functional language design.
+
+There are two broad kinds of lambda calculus: _untyped_ and _typed_.
```
The introduction feels a bit too confused and a tad bit too long. 

Perhaps you want to just mention that it is at the bottom of many functional programming languages and used extensively, and add a "Where is lambda calculus used?" section at the end?

I think this would improve readability (and SEO as well 😅).

lierdakil replied:
I've cut down the intro a bit (arguably a lot), and moved the list of languages based on λ-calculus to the conclusion. Hopefully this is will work better for SEO :laughing:

Heimdell commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,505 @@
+Formally, beta-reduction uses the following rules.
+Given an expression of the form $(\lambda x. t_1)\; t_2$ (i.e. an application of some abstraction to some expression $t_2$):
+
+1. Rename bound variables in $t_1$ and $t_2$ to avoid ambiguity.
```
Maybe, introduce de Bruijn indices ~and alpha-equivalence here~?

lierdakil replied:
As I said above, de Bruijn indices aren't very helpful when doing napkin math. I do introduce them when they become relevant (i.e. in exercises)

Heimdell commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,505 @@
+Note how in call-by-name we didn't have to compute $((\lambda p. p)\; c)$ argument, because it disappeared when applying.
+On the other hand, we had to compute $((\lambda z. z)\; a)$ twice.
+
+A practically significant variation on call-by-name is _call-by-need_, which is, at least in theory, used in Haskell.
```
I'd call it "reduce outermost redex with sharing" strategy

Heimdell commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,505 @@
+$$pred = \lambda n. fst\; (n\; ss\; zz).$$
+
+With this definition, $pred\;c_0 = c_0,$ which isn't ideal.
+However, we don't have an encoding for negative numbers, so we don't have much of a choice.
```
Strictly speaking, we can _make_ one.

lierdakil replied:
Oh, strictly speaking, we can make anything computable. The point is to define subtraction given this encoding.

Heimdell reviewed (approved):
Very good!

lierdakil committed:
fixup! [PR-1320] Article about untyped lambda calculus

lierdakil committed:
fixup! First editing read

lierdakil committed:
fixup! [PR-1320] Article about untyped lambda calculus

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,489 @@
+Lambda calculus is also used extensively in programming language theory research and functional language design.
+
+There are two broad kinds of lambda calculus: _untyped_ and _typed_.
+The _untyped_ kind, discussed in this article, expresses an unrestricted computation.
```
This needs one or two more sentences, probably one with information like this: 

"the untyped kind doesn't find much use beyond the computability theory.
Still, it's necessary to understand the untyped lambda calculus to meaningfully discuss the typed variety (or, more accurately, varieties)."

lierdakil replied:
Done.

lierdakil committed:
fixup! [PR-1320] Article about untyped lambda calculus

NaeosPsy committed:
Some edits while going through the second read

Haven't really finished, commiting to avoid merge conflicts.

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,492 @@
+
+Variables in the body of an abstraction that are parameters of said abstraction are called _bound variables_.
+Other variables are called _free variables_.
+
```
Can we have some examples here? Since these terms are used quite often, I would make sure they are nailed down.

lierdakil replied:
Added a couple examples. Sorry if the phrasing is a bit awkward, I'm feeling a little under the weather today, so words come a little harder than usual.

lierdakil commented on untyped-lambda-calculus/article.md:
```diff
@@ -148,19 +162,6 @@ A practically significant variation on call-by-name is _call-by-need_, which is,
-
-</details>
-</p>
-
```
@NaeosPsy, ?

Did you mean to move it somewhere, or did you decide this is irrelevant?

NaeosPsy replied:
I did decide it is not very relevant. It also looked very ugly when rendered so I thought it would be better if it was gone.

lierdakil replied:
Well, it's arguably an important caveat to keep in mind, but I guess it's fine to remove, as it is only tangentially related to the topic at hand. :ok_hand:

lierdakil committed:
fixup! [PR-1320] Article about untyped lambda calculus

NaeosPsy committed:
Second edit

Finished the second thru-reading of the article, adjusted introduction and conclusion, a lot of small grammar changes.

NaeosPsy commented on untyped-lambda-calculus/article.md:
Unfortunately, this doesn't go with the list, I assume there needs to be some indentation. 🤔

<img width="526" alt="image" src="https://user-images.githubusercontent.com/35335196/165775232-6862605b-8440-491e-9552-8f9b4eb4c07f.png">

NaeosPsy replied:
```
4. As you might remember, a semigroup is a monoid without the identity element. Can you come up with an example of a structure that is a semigroup but is not a monoid? 

    <details>
      <summary>Solution</summary>
    
	One example is [non-empty lists](https://hackage.haskell.org/package/base-4.16.0.0/docs/Data-List-NonEmpty.html) with the `++` operator. While a regular list can be empty, and the empty list thus satisfies the requirement of an identity element, non-empty lists have no such element.
	
    <hr>
    </details>
```

Here's what worked for my monoid article, if that helps.

lierdakil replied:
My renderer renders `details` in a list item ugly, but maybe it's not an issue on the blog itself. Pushed a fixup.

NaeosPsy commented:
Did the second read, all looks good for me except the evaluation strategy section which still feels a bit too much/complicated/hard-to-read.

lierdakil commented on untyped-lambda-calculus/article.md:
```diff
@@ -124,12 +126,12 @@ Any expression where beta-reduction can be applied, we'll call a redex (short fo
 In the left term, it refers to the bound variable; in the right one, it refers to a free variable.
 
-To avoid ambiguity, we can rename $y$ in the first expression to, say, $z$:
+To avoid ambiguity, we can rename $y$ in the first expression to, say, $z$: 
```
eh, the comment about spurious space was meant to go here.

lierdakil committed:
fixup! [PR-1320] Article about untyped lambda calculus

lierdakil commented:
> the evaluation strategy section which still feels a bit too much/complicated/hard-to-read.

I'll see if I can do something about it later today.

lierdakil committed:
fixup! [PR-1320] Article about untyped lambda calculus

lierdakil commented:
I've tried rewriting the evaluation strategies section. Don't know if that helped at all, but at least I've caught an embarrassing mistake :sweat_smile: (TL;DR: I've called all strategies normalizing, due to a combination of bad phrasing and an unfortunate edit; this is demonstrably false; I'm a bit surprised nobody seemed to notice)

NaeosPsy committed:
Structuring evaluation strategy section

Added subheaders and a table, reworded a few things, and cut a few "notes" from the end. More as a proposal than something set in stone, please say if you don't like something.

lierdakil commented on untyped-lambda-calculus/article.md:
```diff
@@ -146,63 +146,73 @@ This brings us to the discussion of evaluation strategies.
+#### Direction of reducing
+
+We can reduce an expression left-to-right or right-to-left.
+This decision is easy: going right-to-left is strictly less powerful because the process doesn't terminate on some terms whereas going left-to-right does.
```
Okay, my fault, but this doesn't sound right. What this meant to say is probably

> going right-to-left is strictly less powerful because the process doesn't terminate on all terms going left-to-right does, but not vice versa.

NaeosPsy replied:
Updated

lierdakil commented on untyped-lambda-calculus/article.md:
```diff
@@ -146,63 +146,73 @@ This brings us to the discussion of evaluation strategies.
 
-Second, we'll decide whether we're first reducing the outermost or the innermost redex.
-To see what is meant by outermost and innermost, consider the following expression:
+#### Direction of reducing
```
Sounds awkward. Perhaps 

> Direction of reduction

?

lierdakil commented on untyped-lambda-calculus/article.md:
```diff
@@ -146,63 +146,73 @@ This brings us to the discussion of evaluation strategies.
+| Applicative order    | Innermost                       | Strong          |
+| Call-by-name (eager) | Outermost                       | Weak            |
+| Call-by-value (lazy) | Innermost                       | Weak            |
+| Call-by-name         | Innermost                       | Weak            |
```
Call-by-name is mentioned twice. I think you meant call-by-need the second time?

A little strange having it in this table here, as it is entirely unclear what differentiates it from call-by-name. It is also considered "lazy evaluation". Actually, under some definition of "lazy", call-by-need is the laziest of them all.

NaeosPsy replied:
The problem is that since it is mentioned, it will also look odd if it doesn't show up. 😅 

Perhaps we can write Call-by-need (Haskell) 🤔?

lierdakil commented on untyped-lambda-calculus/article.md:
```diff
@@ -146,63 +146,73 @@ This brings us to the discussion of evaluation strategies.
-
-Speaking of reductions, let's also mention _eta_-reduction ($\eta$-reduction).
-The rule of eta-reduction says that $\lambda x. f\;x = f,$ if $x$ does not appear free in $f$.
-It boils down to the statement that two functions are the same if and only if they give the same result for all possible arguments.
```
Okay, I didn't find a good place to mention this (besides here, I mean), but it is important to mention this, because the term η-reduction often gets thrown around when discussing Haskell programs.

lierdakil commented on untyped-lambda-calculus/article.md:
```diff
@@ -146,63 +146,73 @@ This brings us to the discussion of evaluation strategies.
 
-If a given term has a normal form under both strong strategies, it's the same for both.
-Note that applicative order strategy is strictly weaker than normal order strategy.
-Some terms that normalize under normal order don't under applicative order.
```
> Some terms that normalize under normal order don't under applicative order.

This is a pretty important point, actually. It follows from the rest of the article, but I think it's good to spell it out.

At least, it would be good to mention that normal order/call-by-name are respectively strongly/weakly normalizing, whereas other strategies aren't necessarily.

lierdakil commented on untyped-lambda-calculus/article.md:
```diff
@@ -146,63 +146,73 @@ This brings us to the discussion of evaluation strategies.
-Expressions that are not reducible using the chosen evaluation strategy are said to be in the _normal form_.
+To sum up, here's a table with all the mentioned strategies and their characteristics:
 
-If a given term has a normal form under both strong strategies, it's the same for both.
```
> If a given term has a normal form under both strong strategies, it's the same for both.

This is also an important point: the result of the computation doesn't change whether we choose to reduce innermost-first or outermost-first; unless the computation diverges under either, that is.

lierdakil commented on untyped-lambda-calculus/article.md:
```diff
@@ -146,63 +146,73 @@ This brings us to the discussion of evaluation strategies.
+|----------------------|---------------------------------|-----------------|
+| Normal order         | Outermost                       | Strong          |
+| Applicative order    | Innermost                       | Strong          |
+| Call-by-name (eager) | Outermost                       | Weak            |
+| Call-by-value (lazy) | Innermost                       | Weak            |
```
It's the reverse. Call-by-name is lazy and call-by-value is eager.

NaeosPsy replied:
Oh, yes, sorry. 🤦‍♂️

This is a good example of what can happen when you reverse the order of things you mention: some people can't read at all. 😂

NaeosPsy replied:
I've updated the construction to be parallel and for the explanations of lazy and eager to be next to call-by-name and call-by-value. Hope it is clearer that way. And of course, I've fixed the thing you commented on.

lierdakil reviewed (commented):
Okay, I have, uh, let's call it "objections", to the last edit. Let's discuss.

NaeosPsy committed:
Addressing objections

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,549 @@
+
+### Eta-reduction
+
+Speaking of reductions, let's also briefly mention _eta_-reduction ($\eta$-reduction).
```
@lierdakil I put this in a specific subsection. It makes more sense now. You might want to change the text a little, I obviously don't know what I am talking about or why it is relevant to Haskell users. 😂

NaeosPsy commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,549 @@
+
+In other words, the applicative order strategy first reduces all of the arguments (from left to right) and then applies the abstraction, while normal order does the inverse.
+
+Expressions that are not reducible using the chosen evaluation strategy are said to be in the _normal form_.
```
@lierdakil Since it mentions only strong strategies, I decided to put the comparison here. If it is more of a general statement on the superiority of outermost-first reducing, it should be at the end, but then it should also somehow mention them, I think. If it is so, you're welcome to rewrite and put at the end of the strategy section.

lierdakil committed:
fixup! [PR-1320] Article about untyped lambda calculus

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,552 @@
+It's a theoretical framework used to define the meaning of computation in many functional languages, such as Haskell, Agda, Idris, etc.
+Understanding lambda calculus can be very helpful when talking about programs written in these languages.
+
+Perhaps surprisingly, understanding lambda calculus can also be helpful in understanding the C++ template metalanguage since it's a functional language in its own right (although it's not, strictly speaking, based on lambda calculus).
```
The reader might want to figure out how lambda is helpful in C++ template metalanguage. Are there any references?

lierdakil replied:
Well, I can point to (Thielecke. "An introduction to C++ template programming", University of Birmingham, 2016), but that's course notes and are not easy to find.

Beyond that, we have λ-calculus-inspired things in Boost, e.g. [the `lambda` metafunction](https://www.boost.org/doc/libs/1_52_0/libs/mpl/doc/tutorial/the-lambda-metafunction.html). Besides, people have a lot of fun researching the isomorphism between C++ templates and λ-calculus, e.g. (Sinkovics, Porkoláb. "Expressing C++ template metaprograms as lambda expressions." In TFP’09, Jun, pp. 2-4. 2009), (Gozillon, Keir. "Point-Free Templates." In European LLVM Developers Meeting. 2018), and there are a few more, and it is pretty illuminating in its own right (there's a [blog post](https://lettersfromtheoutershell.xyz/2020/10/17/templates-calculus) on this topic too)

Not sure if it makes sense referencing any of those.

NaeosPsy replied:
Yeah, feels like it would be too much.

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,552 @@
+
+Lambda calculus, initially envisioned as a formal logic system, was developed by [Alonzo Church](https://en.wikipedia.org/wiki/Alonzo_Church) around the 1930s to explore the foundations of mathematics.
+
+The initial formulation had a logical inconsistency known as the [Kleene–Rosser paradox].
```
Reference

lierdakil replied:
Eh. So here's the deal: I've added the reference to https://plato.stanford.edu/entries/paradoxes-contemporary-logic/#IncoCertFormLogi, but it was removed in 5690f94 by @NaeosPsy, I'm guessing because it's a bit too opaque for a casual reader.

I don't really care either way, it's not a journal publication where every statement has to be either referenced or proved, but I'd like to avoid editing wars :sweat_smile:

NaeosPsy replied:
You still have a link to Kleene-Rosser paradox, right? I just didn't believe anyone that read this would care enough for two links. 

You can perhaps link the Plato in the Kleene-Rosser paradox hyperlink if it explains the situation better. 

But yeah. I imagine this section to be something that will be mostly quickly glossed over so there is no need to go in-depth.

lierdakil replied:
Okay, let's try this: I've added a journal-style reference in 87137c74d9b2ea5075b6f7d4169037f2b6df4472.

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,552 @@
+
+We introduce new abstractions using the $\lambda$ symbol.
+An abstraction consists of a _head_ and a _body_, separated by the dot ($.$) symbol.
+The head contains the $\lambda$ symbol and the parameter name.
```
I would say 'the argument name'

lierdakil replied:
Fair enough, changed.

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,552 @@
+
+Lambda calculus derives its name from the $\lambda$ symbol used in this notation.
+The symbol was most likely chosen arbitrarily.
+However, there's an apocryphal story that the notation was derived from $\hat x$ morphing first into $\Lambda x$ and later into $\lambda x$.
```
That's not quite arbitrarily. $\hat x$ has been chosen since Church was following Russell's notation from Principia Mathematica, but the editors had no $\hat$ when they were preparing Church's manuscript for publication. You may google that for a good reference.

lierdakil replied:
The reason I'm being wishy-washy here is that I wasn't able to find a good reference for this when I searched a few years back (better than nth-hand rumours and speculation). If you have one, please share. Otherwise, I'm sticking to this version.

NaeosPsy replied:
Perhaps we can just cut the "the symbol was most likely chosen arbitrarily. However,"?

lierdakil replied:
Oh, hey, I did find a better reference. It's a bit inconclusive, though:

> By the way, why did Church choose the notation “λ”?  In [A. Church, 7 July 1964.  Unpublished letter to Harald Dickson] he  stated  clearly  that  it  came  from  the  notation  “x̂”  used  for  class-abstractionby Whitehead and Russell, by first modifying “x̂” to “∧x” to distinguish function-abstraction from class-abstraction, and then changing “∧” to “λ” for ease of printing. This origin was also reported in [J. B. Rosser. Highlights of the history of the lambda calculus.Annalsof the History of Computing, 6:337–349, 1984, p.338].  On the other hand, in his later years Church told two enquirers that the choice was more accidental: a symbol was needed and “λ” just happened to be chosen.

(from History of Lambda-calculus and Combinatory Logic, Felice Cardone, Roger Hindley, 2006)

Also nothing about Church's editor not having a symbol for x̂.

lierdakil replied:
I've reworded the section slightly to align more with "History of λ-calculus" and added a journal-style reference in 87137c74d9b2ea5075b6f7d4169037f2b6df4472

DanielRrr commented on untyped-lambda-calculus/article.md:
```diff
@@ -0,0 +1,552 @@
+Hence, $f\;a\;b\;c$ is $((f\;a)\;b)\;c$, i.e. $f$ is applied to $a$, then the result is applied to $b$, then to $c$.
+Since application binds more tightly, we don't need parentheses around the body of an abstraction.
+However, we might need parenthesis around the whole abstraction itself.
+We will also eschew parentheses for nested abstractions, e.g. $\lambda x. \lambda y. x\;y$ is $\lambda x. (\lambda y. (x\;y)).$
```
The example with both nested abstractions and applications might be helpful as well.

lierdakil replied:
I think that's going to be a bit overkill -- the rules are ultimately pretty simple, and getting too caught up in the examples will break the flow.

DanielRrr reviewed (commented):
Looks great! Here are a couple of tiny remarks

lierdakil committed:
fixup! Addressing objections

lierdakil committed:
fixup! Addressing objections

lierdakil committed:
fixup! Addressing objections

lierdakil committed:
Amend wording on full beta-reduction.
